chasm/lib/callback/executors.go

```diff
@@ -0,0 +1,279 @@
+	"go.temporal.io/server/service/history/queues"
+	"go.uber.org/fx"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+// HTTPCaller is a method that can be used to invoke HTTP requests.
+type HTTPCaller func(*http.Request) (*http.Response, error)
+type HTTPCallerProvider func(queues.NamespaceIDAndDestination) HTTPCaller
+
+func NewInvocationTaskExecutor(opts InvocationTaskExecutorOptions) *InvocationTaskExecutor {
+	return &InvocationTaskExecutor{
+		InvocationTaskExecutorOptions: opts,
+	}
+}
+
+type InvocationTaskExecutorOptions struct {
+	fx.In
+
+	Config             *Config
+	NamespaceRegistry  namespace.Registry
+	MetricsHandler     metrics.Handler
+	Logger             log.Logger
+	HTTPCallerProvider HTTPCallerProvider
+	HTTPTraceProvider  commonnexus.HTTPClientTraceProvider
+	HistoryClient      resource.HistoryClient
+	ChasmEngine        chasm.Engine
+}
+
+type InvocationTaskExecutor struct {
+	InvocationTaskExecutorOptions
+}
+
+func (e InvocationTaskExecutor) Execute(ctx context.Context, ref chasm.ComponentRef, attrs chasm.TaskAttributes, task *callbackspb.InvocationTask) error {
+	return e.executeInvocationTask(ctx, ref, attrs, task)
+}
+
+func (e InvocationTaskExecutor) Validate(ctx chasm.Context, cb Callback, attrs chasm.TaskAttributes, task *callbackspb.InvocationTask) (bool, error) {
+	return cb.Attempt == task.Attempt && cb.Status == callbackspb.CALLBACK_STATUS_SCHEDULED, nil
+}
+
+// invocationResult is a marker for the callbackInvokable.Invoke result to indicate to the executor how to handle the
+// invocation outcome.
+type invocationResult interface {
+	// A marker for all possible implementations.
+	mustImplementInvocationResult()
+	error() error
+}
+
+// invocationResultFail marks an invocation as successful.
+type invocationResultOK struct{}
+
+func (invocationResultOK) mustImplementInvocationResult() {}
+
+func (invocationResultOK) error() error {
+	return nil
+}
+
+// invocationResultFail marks an invocation as permanently failed.
+type invocationResultFail struct {
+	err error
+}
+
+func (invocationResultFail) mustImplementInvocationResult() {}
+
+func (r invocationResultFail) error() error {
+	return r.err
+}
+
+// invocationResultRetry marks an invocation as failed with the intent to retry.
+type invocationResultRetry struct {
+	err error
+}
+
+func (invocationResultRetry) mustImplementInvocationResult() {}
+
+func (r invocationResultRetry) error() error {
+	return r.err
+}
+
+type callbackInvokable interface {
+	// Invoke executes the callback logic and returns the invocation result.
+	Invoke(ctx context.Context, ns *namespace.Namespace, e InvocationTaskExecutor, task *callbackspb.InvocationTask) invocationResult
+	// WrapError provides each variant the opportunity to wrap the error returned by the task executor for, e.g. to
+	// trigger the circuit breaker.
+	WrapError(result invocationResult, err error) error
+}
+
+func (e InvocationTaskExecutor) executeInvocationTask(
```

Comment:
You're going to want to call this `Execute` but you'll figure that out when you register it in the library.